Initial event-native documentation (#935)
gravitee-io/gravitee-docs: pages/apim/3.x/event-native-apim/event-native-apim-documentation.adoc

[[event-native-apim-documentation]]
= Documentation
:page-sidebar: apim_3_x_sidebar
:page-permalink: apim/3.x/event_native_apim_documentation.html
:page-folder: apim/event-native-apim
:page-layout: apim3x

[label label-version]#New in version 3.20.0#
[label label-version]#BETA release#

== Overview

_This page is for the draft event-native APIM docmentation that Andrew is writing for 3.20. It is here only for initial draft and review purposes. It will be split into different pages and (probably) moved to a new location later._

== Asynchronous APIs

> In this release, you can use the gateway to connect to asynchrous APIs. You can configure asynchronous API connections using the RESTful interface to the management API.

Reviewer: asynchrous -> asynchronous?

Author: That's the deliberate typo to see if you are paying attention.... ;)

> The OpenAPI speccification for the management API is available at link:/apim/3.x/management-api/3.19/swagger.json[`swagger.json`] (link:/apim/3.x/apim_installguide_rest_apis_documentation.html#apim_console_api_reference[reference documentation]). Examples of how to use the management API to configure asynchrous APIs can be found in the link:https://github.com/gravitee-io/postman-collections[Gravitee Postman Collections], which are described in link:event_native_apim_using.html[Using Event-native API Management].

Reviewer: speccification -> specification?

The following entrypoints are available.

* SSE
* Websockets
* HTTP GET
* HTTP POST

The following endpoints are available

* Kafka

> * MQQT

Reviewer: MQTT

> * MQQT

Reviewer 2: MQTT

== Create a v4 async API 

> The link:https://github.com/gravitee-io/postman-collections[Gravitee Postman Collections] contain several examples of how to create an API.

Reviewer 3: This Github repository is not public for now. Should we consider moving that one from private to public ?

Author: Yes, this will be made available on Postman and the link has been updated.

                    "configuration": {
                        "requestHeadersToMessage": false
                    }
                }
            ]
        }
    ],
    "endpointGroups": [
        {
            "name": "default",
            "type": "kafka",
            "endpoints": [
                {
                    "name": "default",
                    "type": "kafka",
                    "weight": 1,
                    "inheritConfiguration": false,
                    "configuration": {
                        "bootstrapServers": "kafka:9092",
                        "topics" : ["demo"],
                        "consumer" : {
                            "enabled": false
                        },
                        "producer": {
                            "enabled": true
                        }
                    }
                }
            ]
        }
    ],
    "flows": [
        {
            "name": "",
            "selectors": [],
            "request": [],
            "response": [],
            "subscribe": [],
            "publish": [],
            "enabled": true
        }
    ]
}
----

== Quality of Service 

> You can set quality of service with the `qos` object in the `entrypoints` object, as shown in the following example. It can have a value of `none`, `auto`, `at-most-once`, or `at-least-once`.

Reviewer 2: I'm not a big fan of copy-paste of the above descriptor as there is 100% chances that it will go out of sync (and probably wrong) in the next couple of weeks.

> You can set quality of service with the `qos` object in the `entrypoints` object, as shown in the following example. It can have a value of `none`, `auto`, `at-most-once`, or `at-least-once`.

Author: > As mentioned, there is a compatibility matrice to display for QoS capabilities between entrypoint & endpoint

Where do I find the compatibility matrix?

> You can set quality of service with the `qos` object in the `entrypoints` object, as shown in the following example. It can have a value of `none`, `auto`, `at-most-once`, or `at-least-once`.

Author: > I'm not a big fan of copy-paste of the above descriptor as there is 100% chances that it will go out of sync (and probably wrong) in the next couple of weeks.

Would it be better to say something like "Refer to the JSON definition of the API for the values that `qos` can have" and add a link to the JSON?

> You can set quality of service with the `qos` object in the `entrypoints` object, as shown in the following example. It can have a value of `none`, `auto`, `at-most-once`, or `at-least-once`.

Author: Compatibility matrixes added and reference to the JSON definition added.

[source json]
----
"entrypoints": [
                {
                    "type": "sse",
                    "qos": "none",
                    "configuration": {
                        "heartbeatIntervalInMs": 5000,
                        "metadataAsComment": false,
                        "headersAsComment": false
                    }
                }
            ]
----

== Policies

> You can set a policy on a subscription by setting the `policy` object in the `subscribe` object, as shown in the following example.

Author: I have updated the documentation to make it more general.

[[event-native-apim-documentation]]
= Documentation
:page-sidebar: apim_3_x_sidebar
:page-permalink: apim/3.x/event_native_apim_documentation.html
:page-folder: apim/event-native-apim
:page-layout: apim3x

[label label-version]#New in version 3.20.0#
[label label-version]#BETA release#

[NOTE]
====
This document assumes you are familiar with synchronous APIs, asynchronous APIs, and the OpenAPI specification.
====

[TIP]
====
Before running any of the examples on this page, work through some of the  link:event_native_apim_example_use_cases/html[example use cases] so you know that the event-native engine is running.
====

== Asynchronous APIs

> In this release, you can use the gateway to connect to asynchronous APIs. You can configure asynchronous API connections using the RESTful interface to the management API.

Author: OK, changed.

> The OpenAPI specification for the management API is available at link:/apim/3.x/management-api/3.19/swagger.json[`swagger.json`] (link:/apim/3.x/apim_installguide_rest_apis_documentation.html#apim_console_api_reference[reference documentation]). Examples of how to use the management API to configure asynchronous APIs can be found in the link:https://github.com/gravitee-io/postman-collections[Gravitee Postman Collections], which are described in link:event_native_apim_example_use_cases.html[example use cases].

Author: That's covered below.

[NOTE]
====
The management API is in beta and it may change.
====

== Entrypoints and endpoints

> An _entrypoint_ is how an end user connects to the Gravitee gateway. An _endpoint_ is how the Gravitee gateway connects to the upstream API.

Author: Thanks. I've made this change.

> The following types of entrypoint are available.

Reviewer 4: We need a proper explanation for each entry points, what they do, how they work, what they achieve.

Author: Link added for now, and will expand this section in the next iteration.

* SSE
* SSE advanced
* Websockets
* HTTP GET
* HTTP POST
* Webhook
* Webhook advanced (EE)

The following types of endpoint are available

* Kafka
* Kafka advanced (EE)
* MQTT
* MQTT advanced (EE)

> == Create a v4 async API 

Reviewer 4: This is for me too technical and straighforward. We need to explain what have changed, what is a listener, what is an entrypoint, what is an endpoint, what is the new flow, how they interact

                    "weight": 1,
                    "inheritConfiguration": false,
                    "configuration": {
                        "bootstrapServers": "kafka:9092",
                        "topics" : ["demo"],
                        "consumer" : {
                            "enabled": false
                        },
                        "producer": {
                            "enabled": true
                        }
                    }
                }
            ]
        }
    ],
    "flows": [
        {
            "name": "",
            "selectors": [],
            "request": [],
            "response": [],
            "subscribe": [],
            "publish": [],
            "enabled": true
        }
    ]
}
----

== Quality of Service 

When working with asynchronous APIs, quality of service is important. Quality of service defines the guaranteed level of message delivery. For example, a quality of service of "none" means that a given message might be delivered zero, one, or several times. A quality of service of "at-most-once" means that a given message will be delivered zero or one times, with no duplication.

> A higher quality of service leads to lower system performance.

Reviewer 4: ```suggestion
A higher quality of service could lead to lower system performance depending on the endpoint choosen.
```

[source json]
----
"entrypoints": [
                {
                    "type": "sse",
                    "qos": "none",
                    "configuration": {
                        "heartbeatIntervalInMs": 5000,
                        "metadataAsComment": false,
                        "headersAsComment": false
                    }
                }
            ]
----

=== Compatibility

Not all qualities of service work with every combination of entrypoint and endpoint. The following table shows how they can be combined.

.Table Quality of service compatibility matrix
|===
| Entrypoint| MQTT5 | MQTT5 advanced | Kafka | Kafka advanced

| SSE
| none, auto
| none, auto
| none, auto
| none, auto

| SSE advanced
| none, auto
| none, auto
| none, auto
| none, auto, at least once, at most once

| HTTP POST
| none, auto
| none, auto
| none, auto
| none, auto

| HTTP GET
| auto
| auto
| auto
| auto, at least once, at most once

| Websocket
| none, auto
| none, auto
| none, auto
| none, auto

| SSE advanced
| none, auto
| none, auto, at least once, at most once
| none, auto
| none, auto, at least once, at most once

|===

== Policies

> Policies are steps in the gateway proxy chain. A policy guarantees that a given business rule with be fulfilled during processing.

Reviewer 4: ```suggestion
Policies are steps in the gateway execution chain. A policy guarantees that a given business rule with be fulfilled during processing.
```

> Policies can be set on requests, responses, subscriptions, and messages. The following example shows how to set a policy on a subscription.

Reviewer 4: ```suggestion
Policies can be set on request, response, subscribe, or publish phases. The following example shows how to set a policy on a subscribe phase.
```

[source json]
----
"subscribe": [
                {
                    "name": "Message filtering",
                    "description": "Apply filter to messages",
                    "enabled": true,
                    "policy": "message-filtering",
                    "configuration": {
                        "filter": "{#message.headers.foo == #subscription.metadata['bar']}"
                    }
                }
            ]
----

For an example, see _04 - Event Consumption - Webhook_ > _Webhook Messaging Filtering_ > _Create API_ in The link:https://github.com/gravitee-io/postman-collections[Gravitee Postman Collections].

== Consume a v4 API

> The link:https://github.com/gravitee-io/postman-collections[Gravitee Postman Collections] contain several examples of how to consume an API. Some examples are described on link:event_native_apim_example_use_cases.html#event_consumption[Event consumption].

Reviewer 4: I don't understand what you mean here.

Author: OK, paragraph updated.

[[event-native-apim-documentation]]
= Documentation
:page-sidebar: apim_3_x_sidebar
:page-permalink: apim/3.x/event_native_apim_documentation.html
:page-folder: apim/event-native-apim
:page-layout: apim3x

[label label-version]#New in version 3.20.0#
[label label-version]#BETA release#

== Overview

== Asynchronous APIs

In this release, you can use the gateway to connect to asynchronous APIs. You can configure asynchronous API connections using the RESTful interface to the management API.

The OpenAPI specification for the management API is available at link:/apim/3.x/management-api/3.19/swagger.json[`swagger.json`] (link:/apim/3.x/apim_installguide_rest_apis_documentation.html#apim_console_api_reference[reference documentation]). Examples of how to use the management API to configure asynchronous APIs can be found in the link:https://github.com/gravitee-io/postman-collections[Gravitee Postman Collections], which are described in link:event_native_apim_using.html[Using Event-native API Management].

[NOTE]
====
The management API is in beta and it may change.
====

> == Entrypoints and endpoints

Author: This is good information, and I will integrate it into the documentation next year.

                    "configuration": {
                        "requestHeadersToMessage": false
                    }
                }
            ]
        }
    ],
    "endpointGroups": [
        {
            "name": "default",
            "type": "kafka",
            "endpoints": [
                {
                    "name": "default",
                    "type": "kafka",
                    "weight": 1,
                    "inheritConfiguration": false,
                    "configuration": {
                        "bootstrapServers": "kafka:9092",
                        "topics" : ["demo"],
                        "consumer" : {
                            "enabled": false
                        },
                        "producer": {
                            "enabled": true
                        }
                    }
                }
            ]
        }
    ],
    "flows": [
        {
            "name": "",
            "selectors": [],
            "request": [],
            "response": [],
            "subscribe": [],
            "publish": [],
            "enabled": true
        }
    ]
}
----

== Quality of Service 

> You can set quality of service with the `qos` object in the `entrypoints` object, as shown in the following example. See the link:/apim/3.x/management-api/3.19/swagger.json[`swagger.json`] definition of the Management API For values that `qos` can have.

Author: I think what we have at the moment is adequate for this release of the documentation, but I will expand it in the next iteration.
